pkg/stacks/truncate/truncate.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package truncate provides functions for truncating Kubernetes values in a
+// predictable way offering mildly collision safe values usable in
+// deterministic field searches.
+package truncate
+
+import (
+	// sha1 is not cryptographically secure, but that is not a goal. we require
+	// predictable and uniform text transformation. Any checksum function with
+	// even distribution would suffice.
+	"crypto/sha1" // nolint:blacklist
+	"encoding/base32"
+	"fmt"
+	"strings"
+)
+
+const (
+	// LabelNameLength is the max length of a name fragment in a label
+	LabelNameLength = 63
+
+	// LabelValueLength is the max length of a label value
+	LabelValueLength = 63
+
+	// ResourceNameLength is the max length of a resource name or a label with
+	// optional prefix and name
+	ResourceNameLength = 253
+
+	// DefaultSuffixLength is the length of combined separator and checksum
+	// characters that will be used as a truncation suffix in Label and Resource
+	// truncation functions
+	DefaultSuffixLength = 6
+
+	sha1Length = 40
+)
+
+// Truncate replaces the suffixLength number of trailing characters from str
+// with a consistent hash fragment based on that string. The suffix will include
+// a leading hyphen.
+// An error will be returned if the truncation length is less than the
+// suffixLength, or truncation length is greater than sha1Length, or the suffix
+// length is less than 2.
+func Truncate(str string, length, suffixLength int) (string, error) {
```

Comment:
I have a really dumb question. Did we already consider using or copying Kubernetes's implementation of this same behavior (that they use for pods, for example)? It seems like maybe the implementation is not long enough to be worth it.

Reply:
Like Kubernetes, we are using labels to identify the related resources (matchLabels in a deployment refer to the labels in the pod spec).

Our resource names are extra long because we are taking two 253 characters strings (tenant namespace and tenant resource name) and appending them into at most a 253 character slot (the host resource name).  Compounding this further, we need to track these potentially 506 character names as label values, which can be no longer than 63 characters. 

Kubernetes does not offer a solution for this. If you try to create a deployment with a name > 63 characters, the admission controller will reject it.  To apply that same resolution here, if the namespace and name of the tenant Stack resource exceeded 62 characters (losing one for a separator), it would need to be rejected.

This PR seeks to fix the current implementation and avoid changing the existing behavior. Since resources with long name concatenations are not currently supported, this PR should not affect existing resources with the new truncation behavior.  A broader change to how resource names are stored on the host could result in orphaned resources in existing host environments.

Annotations offer sufficient length to store the tenant resource names on the host resource without modification.  The tenant resource names are being stored on the host for debug purposes only.

Reply:
You may have been referring to the Kubernetes suffix pattern that uses random characters as a suffix (deployment `foo`, replica scale 2 resulting in pod `foo-abc` and `foo-zyx`). 

In order to provide host resource names that could be determined given tenant resource names, randomization was avoided.

Could tenant resource labels have been used to to fetch host resources? Perhaps, but existing resources are not labeled.
Are host resource labels (referring to the tenant) sufficient? Perhaps, but existing resources are not labeled.
Are host resource annotations sufficient? They are not indexed, so a query to find a matching deployment on a host, given a tenant Stack name would need to query all stacks (from all namespaces in a tenant cluster).  Again, these annotations are not present on existing resources.